### PR TITLE
selectFromResult for mutations

### PR DESCRIPTION
**Pull request by msutkowski**
- Adds `selectFromResult` option for `useMutation()`. 
  - This allows a user to 'optimize renders' by _only_ rerendering when the selected values change. There will always be 1 rerender when the mutation is called as it creates a new `requestId`. For the most performant implementation, you can `yourEndpoint.useMutation(undefined, { selectFromResult: () => ({}) })`
  - All possible usages can be seen in the tests with the appropriate assertions :sweat_smile:

### CODE REVIEW

**Review by markerikson — APPROVED** [review]

**Review by msutkowski — COMMENTED** [review]
This change will require updates in the docs that were merged: https://github.com/rtk-incubator/rtk-query/pull/191/files#diff-b2cf37354c870ebf35a4045446fae1006ca83193784282bb2066da6bba8a1f91R94, but was introduced for consistency with the other hook names

File: src/react-hooks/buildHooks.ts [hunk 1]
```diff
@@ -423,8 +474,18 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
...
+          createSelector([select(requestId || skipSelector)], (subState) =>
+            selectFromResult(subState, defaultMutationStateSelector)
+          ),
+        [select, requestId, selectFromResult]
+      );
+
+      const currentState = useSelector(
+        (state: RootState<Definitions, any, any>) => mutationSelector(state),
+        shallowEqual
+      );
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:488]:
```suggestion
      const currentState = useSelector(mutationSelector, shallowEqual);
```

File: src/react-hooks/buildHooks.ts [hunk 2]
```diff
@@ -128,15 +127,67 @@ type UseQueryStateDefaultResult<D extends QueryDefinition<any, any, any, any>> =
...
 
-export type MutationHook<D extends MutationDefinition<any, any, any, any>> = () => [
+export type MutationStateSelector<R, D extends MutationDefinition<any, any, any, any>> = (
+  state: MutationResultSelectorResult<D>,
+  defaultMutationStateSelector: DefaultMutationStateSelector<D>
+) => R;
+
+export type DefaultMutationStateSelector<D extends MutationDefinition<any, any, any, any>> = (
+  state: MutationResultSelectorResult<D>
+) => UseMutationStateDefaultResult<D>;
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:137]:
```suggestion
) => MutationResultSelectorResult<D>;
```

File: src/react-hooks/buildHooks.ts [hunk 3]
```diff
@@ -128,15 +127,67 @@ type UseQueryStateDefaultResult<D extends QueryDefinition<any, any, any, any>> =
...
+  | Override<Extract<UseMutationStateBaseResult<D>, { status: QueryStatus.uninitialized }>, { isUninitialized: true }>
+  | Override<
+      UseMutationStateBaseResult<D>,
+      | { isLoading: true; data: undefined }
+      | ({ isSuccess: true; error: undefined } & Required<
+          Pick<UseMutationStateBaseResult<D>, 'data' | 'fulfilledTimeStamp'>
+        >)
+      | ({ isError: true } & Required<Pick<UseMutationStateBaseResult<D>, 'error'>>)
+    >
+>;
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:174]:
```suggestion
```

File: src/react-hooks/buildHooks.ts [hunk 4]
```diff
@@ -128,15 +127,67 @@ type UseQueryStateDefaultResult<D extends QueryDefinition<any, any, any, any>> =
...
+
+export type UseMutation<D extends MutationDefinition<any, any, any, any>> = <R = UseMutationStateDefaultResult<D>>(
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:176]:
```suggestion
export type UseMutation<D extends MutationDefinition<any, any, any, any>> = <R = MutationResultSelectorResult<D>>(
```

File: src/react-hooks/buildHooks.ts [hunk 5]
```diff
@@ -128,15 +127,67 @@ type UseQueryStateDefaultResult<D extends QueryDefinition<any, any, any, any>> =
...
     arg: QueryArgFrom<D>
   ) => {
     unwrap: () => Promise<ResultTypeFrom<D>>;
   },
-  MutationSubState<D> & RequestStatusFlags
+  UseMutationStateResult<D, R>
 ];
 
+const defaultMutationStateSelector: DefaultMutationStateSelector<any> = (currentState) => {
+  return currentState as UseMutationStateDefaultResult<any>;
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:188]:
```suggestion
  return currentState;
```

File: test/buildHooks.test.tsx [hunk 6]
```diff
@@ -1301,4 +1301,130 @@ describe('hooks with createApi defaults set', () => {
...
+      }),
+    });
+
+    const storeRef = setupApiStore(defaultApi, {
+      ...actionsReducer,
+    });
+
+    let getRenderCount: () => number = () => 0;
+
+    it('does not cause a rerender when using selectFromResult with an empty object', async () => {
```
Comment (phryneas) [inline test/buildHooks.test.tsx:1333]:
```suggestion
    it('causes no more than one rerender when using selectFromResult with an empty object', async () => {
```

**Review by phryneas — CHANGES_REQUESTED** [review]
Already said it in chat - I think we don't need the fancy type overloading here as we don't really have a use for `isFetching` in mutations.

My change suggestions might be a bit sloppy, I'm not feeling so well :upside_down_face:

File: src/core/apiState.ts [hunk 7]
```diff
@@ -132,9 +132,9 @@ export type MutationSubState<D extends BaseEndpointDefinition<any, any, any>> =
   | (({
       status: QueryStatus.fulfilled;
     } & WithRequiredProp<BaseMutationSubState<D>, 'data' | 'fulfilledTimeStamp'>) & { error: undefined })
-  | ({
+  | (({
       status: QueryStatus.pending;
-    } & BaseMutationSubState<D>)
+    } & BaseMutationSubState<D>) & { data?: undefined })
```
Comment (msutkowski) [inline src/core/apiState.ts:137]: Forces the return type of `result.data` from `string | undefined` to `undefined` in the unionTypes test, which is what we want. `data` should never be a string there as a new request is `{ isLoading: true, data: undefined, error: undefined `

File: src/core/buildSelectors.ts [hunk 8]
```diff
@@ -53,6 +53,10 @@ type MutationResultSelector<Definition extends MutationDefinition<any, any, any,
   requestId: string | typeof skipSelector
 ) => (state: RootState) => MutationSubState<Definition> & RequestStatusFlags;
```
Comment (phryneas) [inline src/core/buildSelectors.ts:54]:
```suggestion
) => (state: RootState) => MutationResultSelectorResult<Definition>;
```

File: src/react-hooks/buildHooks.ts [hunk 9]
```diff
@@ -128,15 +120,36 @@ type UseQueryStateDefaultResult<D extends QueryDefinition<any, any, any, any>> =
...
   ) => {
     unwrap: () => Promise<ResultTypeFrom<D>>;
   },
-  MutationSubState<D> & RequestStatusFlags
+  UseMutationStateResult<D, R>
 ];
 
+const defaultMutationStateSelector: DefaultMutationStateSelector<any> = (currentState) => {
+  return currentState;
+};
```
Comment (phryneas) [inline src/react-hooks/buildHooks.ts:151]:
```suggestion
const defaultMutationStateSelector: DefaultMutationStateSelector<any> = (currentState) => currentState;
```

**Review by phryneas — CHANGES_REQUESTED** [review]
mini nitpicks, then this can go in :)

**Pull request merged**